File: crates/kuben-store/src/repo/support.rs
//! What a support bundle says about the database (M4.11): counts and codes
//! only — no names, payloads, secrets or personal data.

use kuben_core::time::now_ms;
use serde::Serialize;

use crate::{Store, StoreError};

/// How far back the bundle looks at operations.
pub const SUPPORT_WINDOW_MS: i64 = 7 * 24 * 3_600_000;

const OPERATIONS: &str = "SELECT kind, phase, coalesce(error_code, '') AS code, count(*) AS n \
     FROM operations WHERE requested_at >= $1 OR NOT done \
     GROUP BY kind, phase, error_code ORDER BY kind, phase, error_code LIMIT 500";
const STUCK: &str = "SELECT count(*) FROM operations \
     WHERE NOT done AND last_progress_at IS NOT NULL AND last_progress_at < $1";
const OUTBOX: &str = "SELECT count(*) FILTER (WHERE delivered_at IS NULL), \
     coalesce(min(created_at) FILTER (WHERE delivered_at IS NULL), 0) FROM outbox";
const DELIVERIES: &str = "SELECT status, count(*) FROM webhook_deliveries \
     WHERE created_at >= $1 GROUP BY status ORDER BY status";
const TOTALS: &str = "SELECT (SELECT count(*) FROM organizations), (SELECT count(*) FROM users), \
     (SELECT count(*) FROM deployment_runs), (SELECT count(*) FROM releases)";

/// Operations of one kind, phase and error code.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, sqlx::FromRow)]
pub struct OperationCount {
    pub kind: String,
    pub phase: String,
    pub code: String,
    pub n: i64,
}

/// The database part of a support bundle.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize)]
pub struct SupportSummary {
    pub organizations: i64,
    pub users: i64,
    pub deployment_runs: i64,
    pub releases: i64,
    pub projects: i64,
    pub environments: i64,
    pub apps: i64,
    pub paused_apps: i64,
    /// Operations requested in the window, and every unfinished one.
    pub operations: Vec<OperationCount>,
    /// Unfinished operations without progress for an hour.
    pub stuck_operations: i64,
    pub outbox_pending: i64,
    /// When the oldest undelivered outbox message was written (0: none).
    pub outbox_oldest_at: i64,
    /// Webhook deliveries of the window by status.
    pub webhook_deliveries: Vec<(String, i64)>,
    /// Open incidents by kind and severity.
    pub open_incidents: Vec<(String, String, i64)>,
    pub disabled_webhooks: i64,
}

const TENANT_COUNTS: &str = "SELECT (SELECT count(*) FROM projects WHERE org_id = $1), \
     (SELECT count(*) FROM environments WHERE org_id = $1), \
     (SELECT count(*) FROM application_targets WHERE org_id = $1), \
     (SELECT count(*) FROM application_targets WHERE org_id = $1 AND paused_at IS NOT NULL), \
     (SELECT count(*) FROM webhook_endpoints WHERE org_id = $1 AND disabled_at IS NOT NULL)";
const TENANT_INCIDENTS: &str = "SELECT kind, severity, count(*) FROM incidents \
     WHERE org_id = $1 AND resolved_at IS NULL GROUP BY kind, severity";

impl Store {
    /// Counts for a support bundle.
    pub async fn support_summary(&self) -> Result<SupportSummary, StoreError> {
        let now = now_ms();
        let (organizations, users, deployment_runs, releases): (i64, i64, i64, i64) =
            sqlx::query_as(TOTALS).fetch_one(self.pool()).await?;
        let (outbox_pending, outbox_oldest_at): (i64, i64) =
            sqlx::query_as(OUTBOX).fetch_one(self.pool()).await?;
        let mut summary = SupportSummary {
            organizations,
            users,
            deployment_runs,
            releases,
            operations: sqlx::query_as(OPERATIONS)
                .bind(now - SUPPORT_WINDOW_MS)
                .fetch_all(self.pool())
                .await?,
            stuck_operations: sqlx::query_scalar(STUCK)
                .bind(now - 3_600_000)
                .fetch_one(self.pool())
                .await?,
            outbox_pending,
            outbox_oldest_at,
            webhook_deliveries: sqlx::query_as(DELIVERIES)
                .bind(now - SUPPORT_WINDOW_MS)
                .fetch_all(self.pool())
                .await?,
            ..SupportSummary::default()
        };
        let mut incidents = std::collections::BTreeMap::<(String, String), i64>::new();
        for org in self.org_ids().await? {
            let mut tenant = self.tenant(org).await?;
            let (projects, environments, apps, paused, disabled): (i64, i64, i64, i64, i64) =
                sqlx::query_as(TENANT_COUNTS)
                    .bind(org.to_string())
                    .fetch_one(&mut *tenant.tx)
                    .await?;
            summary.projects += projects;
            summary.environments += environments;
            summary.apps += apps;
            summary.paused_apps += paused;
            summary.disabled_webhooks += disabled;
            let open: Vec<(String, String, i64)> = sqlx::query_as(TENANT_INCIDENTS)
                .bind(org.to_string())
                .fetch_all(&mut *tenant.tx)
                .await?;
            for (kind, severity, n) in open {
                *incidents.entry((kind, severity)).or_default() += n;
            }
        }
        summary.open_incidents = incidents.into_iter().map(|((k, s), n)| (k, s, n)).collect();
        Ok(summary)
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        repo::NewIncident,
        testing::{pg_store, skip},
    };

    #[tokio::test]
    async fn the_summary_counts_across_organizations() {
        let Some(store) = pg_store().await else {
            return skip("the_summary_counts_across_organizations");
        };
        for slug in ["a", "b"] {
            let org = store.create_org(slug, slug).await.expect("org").id;
            let mut t = store.tenant(org).await.expect("tenant");
            t.create_project("shop", "Shop").await.expect("project");
            let incident = NewIncident {
                project: None,
                environment: None,
                target: None,
                kind: "backup.stale".into(),
                severity: "critical",
                dedupe_key: "k".into(),
                title: "t".into(),
                detail: None,
            };
            t.open_incident(&incident).await.expect("incident");
            t.commit().await.expect("commit");
        }
        let summary = store.support_summary().await.expect("summary");
        assert_eq!((summary.organizations, summary.projects), (2, 2));
        assert_eq!(
            summary.open_incidents,
            [("backup.stale".to_owned(), "critical".to_owned(), 2)]
        );
        assert_eq!(summary.outbox_pending, 0);
        let text = serde_json::to_string(&summary).expect("json");
        assert!(!text.contains("shop"), "no names: {text}");
    }
}

File: internal/store/support.go
package store

// What a support bundle says about the database (M4.11): counts and codes
// only — no names, payloads, secrets or personal data; the port of
// repo/support.rs.

import (
	"cmp"
	"context"
	"encoding/json"
	"errors"
	"slices"

	"github.com/jackc/pgx/v5"

	"github.com/Teamtem-dev/kuben/internal/core/ids"
)

// SupportWindowMs is how far back the bundle looks at operations.
const SupportWindowMs int64 = 7 * 24 * 3_600_000

const (
	supportOperations = "SELECT kind, phase, coalesce(error_code, '') AS code, count(*) AS n " +
		"FROM operations WHERE requested_at >= $1 OR NOT done " +
		"GROUP BY kind, phase, error_code ORDER BY kind, phase, error_code LIMIT 500"
	supportStuck = "SELECT count(*) FROM operations " +
		"WHERE NOT done AND last_progress_at IS NOT NULL AND last_progress_at < $1"
	supportOutbox = "SELECT count(*) FILTER (WHERE delivered_at IS NULL), " +
		"coalesce(min(created_at) FILTER (WHERE delivered_at IS NULL), 0) FROM outbox"
	supportDeliveries = "SELECT status, count(*) FROM webhook_deliveries " +
		"WHERE created_at >= $1 GROUP BY status ORDER BY status"
	supportTotals = "SELECT (SELECT count(*) FROM organizations), (SELECT count(*) FROM users), " +
		"(SELECT count(*) FROM deployment_runs), (SELECT count(*) FROM releases)"
	supportTenantCounts = "SELECT (SELECT count(*) FROM projects WHERE org_id = $1), " +
		"(SELECT count(*) FROM environments WHERE org_id = $1), " +
		"(SELECT count(*) FROM application_targets WHERE org_id = $1), " +
		"(SELECT count(*) FROM application_targets WHERE org_id = $1 AND paused_at IS NOT NULL), " +
		"(SELECT count(*) FROM webhook_endpoints WHERE org_id = $1 AND disabled_at IS NOT NULL)"
	supportTenantIncidents = "SELECT kind, severity, count(*) FROM incidents " +
		"WHERE org_id = $1 AND resolved_at IS NULL GROUP BY kind, severity"
)

// stuckAfterMs is how long an unfinished operation may go without progress
// before the bundle counts it as stuck.
const stuckAfterMs = 3_600_000

// OperationCount is the operations of one kind, phase and error code.
type OperationCount struct {
	Kind  string `json:"kind"`
	Phase string `json:"phase"`
	Code  string `json:"code"`
	N     int64  `json:"n"`
}

// StatusCount is the webhook deliveries of one status. Its JSON is Rust's
// tuple: `[status, n]`.
type StatusCount struct {
	Status string
	N      int64
}

// MarshalJSON writes the pair as an array.
func (c StatusCount) MarshalJSON() ([]byte, error) { return json.Marshal([]any{c.Status, c.N}) }

// IncidentCount is the open incidents of one kind and severity. Its JSON is
// Rust's tuple: `[kind, severity, n]`.
type IncidentCount struct {
	Kind     string
	Severity string
	N        int64
}

// MarshalJSON writes the triple as an array.
func (c IncidentCount) MarshalJSON() ([]byte, error) {
	return json.Marshal([]any{c.Kind, c.Severity, c.N})
}

// SupportSummary is the database part of a support bundle.
type SupportSummary struct {
	Organizations  int64 `json:"organizations"`
	Users          int64 `json:"users"`
	DeploymentRuns int64 `json:"deployment_runs"`
	Releases       int64 `json:"releases"`
	Projects       int64 `json:"projects"`
	Environments   int64 `json:"environments"`
	Apps           int64 `json:"apps"`
	PausedApps     int64 `json:"paused_apps"`
	// Operations requested in the window, and every unfinished one.
	Operations []OperationCount `json:"operations"`
	// StuckOperations are unfinished operations without progress for an
	// hour.
	StuckOperations int64 `json:"stuck_operations"`
	OutboxPending   int64 `json:"outbox_pending"`
	// OutboxOldestAt is when the oldest undelivered outbox message was
	// written (0: none).
	OutboxOldestAt int64 `json:"outbox_oldest_at"`
	// WebhookDeliveries of the window by status.
	WebhookDeliveries []StatusCount `json:"webhook_deliveries"`
	// OpenIncidents by kind and severity.
	OpenIncidents    []IncidentCount `json:"open_incidents"`
	DisabledWebhooks int64           `json:"disabled_webhooks"`
}

// SupportSummary is the counts for a support bundle.
func (s *Store) SupportSummary(ctx context.Context) (SupportSummary, error) {
	const op = "summarize for support"
	now := s.now()
	var sum SupportSummary
	if err := queryOne(ctx, s.db, op, supportTotals,
		[]any{&sum.Organizations, &sum.Users, &sum.DeploymentRuns, &sum.Releases}); err != nil {
		return SupportSummary{}, err
	}
	if err := queryOne(ctx, s.db, op, supportOutbox, []any{&sum.OutboxPending, &sum.OutboxOldestAt}); err != nil {
		return SupportSummary{}, err
	}
	var err error
	sum.Operations, err = queryAll(ctx, s.db, op, supportOperations, func(row pgx.CollectableRow) (OperationCount, error) {
		var c OperationCount
		err := row.Scan(&c.Kind, &c.Phase, &c.Code, &c.N)
		return c, err
	}, now-SupportWindowMs)
	if err != nil {
		return SupportSummary{}, err
	}
	if err := queryOne(ctx, s.db, op, supportStuck, []any{&sum.StuckOperations}, now-stuckAfterMs); err != nil {
		return SupportSummary{}, err
	}
	sum.WebhookDeliveries, err = queryAll(ctx, s.db, op, supportDeliveries, func(row pgx.CollectableRow) (StatusCount, error) {
		var c StatusCount
		err := row.Scan(&c.Status, &c.N)
		return c, err
	}, now-SupportWindowMs)
	if err != nil {
		return SupportSummary{}, err
	}
	orgs, err := s.OrgIDs(ctx)
	if err != nil {
		return SupportSummary{}, err
	}
	incidents := map[[2]string]int64{}
	for _, org := range orgs {
		if err := s.supportTenant(ctx, org, &sum, incidents); err != nil {
			return SupportSummary{}, err
		}
	}
	sum.OpenIncidents = make([]IncidentCount, 0, len(incidents))
	for key, n := range incidents {
		sum.OpenIncidents = append(sum.OpenIncidents, IncidentCount{Kind: key[0], Severity: key[1], N: n})
	}
	// Rust's BTreeMap order: by kind, then severity.
	slices.SortFunc(sum.OpenIncidents, func(a, b IncidentCount) int {
		return cmp.Or(cmp.Compare(a.Kind, b.Kind), cmp.Compare(a.Severity, b.Severity))
	})
	return sum, nil
}

// supportTenant adds org's counts to sum and its open incidents to
// incidents, in one tenant transaction (which Rust never committed: it only
// reads).
func (s *Store) supportTenant(ctx context.Context, org ids.OrgID, sum *SupportSummary, incidents map[[2]string]int64) (err error) {
	const op = "summarize an organization for support"
	t, err := s.Tenant(ctx, org)
	if err != nil {
		return err
	}
	defer func() {
		if rbErr := t.Rollback(ctx); rbErr != nil {
			err = errors.Join(err, rbErr)
		}
	}()
	var projects, environments, apps, paused, disabled int64
	if err := queryOne(ctx, t.tx, op, supportTenantCounts,
		[]any{&projects, &environments, &apps, &paused, &disabled}, org.String()); err != nil {
		return err
	}
	sum.Projects += projects
	sum.Environments += environments
	sum.Apps += apps
	sum.PausedApps += paused
	sum.DisabledWebhooks += disabled
	open, err := queryAll(ctx, t.tx, op, supportTenantIncidents, func(row pgx.CollectableRow) (IncidentCount, error) {
		var c IncidentCount
		err := row.Scan(&c.Kind, &c.Severity, &c.N)
		return c, err
	}, org.String())
	if err != nil {
		return err
	}
	for _, c := range open {
		incidents[[2]string{c.Kind, c.Severity}] += c.N
	}
	return nil
}
